Merge CheckAlpha into texture decoding

hrydgard opened the pull request:
Fixes a recently discovered performance issue with CheckAlpha - on Vulkan and D3D we write decoded texture data directly to video memory, which can be super slow to read from, depending on driver, platform, phase of moon, etc.

This changes the code to "sum up" (AND together) the colors, and then check against the full alpha mask at the end.

Since we can remove CheckAlpha entirely this should improve performance despite the slight extra checking. And even so, it doesn't increase memory traffic, just some extra arithmetic which rarely is the bottleneck in modern out-of-order CPUs.

Disabled the optimization to deswizzle raw textures directly to VRAM, but no big deal as they're not that common and it's pretty fast anyway. Also, I check the input separately in a few cases to avoid writing more special case functions, so there's still room for minor optimizations.

hrydgard committed:
AND together colors while decoding, and then check against fullAlphaMask.

hrydgard committed:
Comments

unknownbrackets commented on GPU/Common/TextureCacheCommon.cpp:
```diff
@@ -1353,7 +1355,10 @@ static inline void ConvertFormatToRGBA8888(GEPaletteFormat format, u32 *dst, con
 template <typename DXTBlock, int n>
-static void DecodeDXTBlock(uint8_t *out, int outPitch, uint32_t texaddr, const uint8_t *texptr, int w, int h, int bufw, bool reverseColors, bool useBGRA) {
+static void DecodeDXTBlocks(uint8_t *out, int outPitch, uint32_t texaddr, const uint8_t *texptr,
+	int w, int h, int bufw, bool reverseColors, bool useBGRA,
+	u32 *alphaSum, u32 *fullAlphaMask) {
```
nit: Maybe we should use a struct.  My general rule of thumb is if I have so many arguments that I'm motivated to put some on another line, then I have too many arguments.

-[Unknown]

hrydgard replied:
Yeah, it's probably time to clean this up.

hrydgard replied:
Removed the two parameters again, good enough for now heh

hrydgard replied:
Not gonna bother for now though.. I got rid of the new parameters from the calling function `DecodeTextureLevel` though.

unknownbrackets commented on GPU/Common/TextureDecoder.h:
```diff
@@ -117,15 +117,23 @@ static const u8 textureBitsPerPixel[16] = {
 	// Usually, there is no special offset, mask, or shift.
 	const bool nakedIndex = gstate.isClutIndexSimple();
 
+	ClutT alphaSum = (ClutT)(-1);
```
Hm, not written anywhere yet.

-[Unknown]

hrydgard committed:
Fix DeIndexTexture

iota97 commented on GPU/Common/TextureCacheCommon.cpp:
```diff
@@ -1366,26 +1371,88 @@ static void DecodeDXTBlock(uint8_t *out, int outPitch, uint32_t texaddr, const u
+		break;
+	}
+
 	w = (w + 3) & ~3;
```
This line happened in: 1f6860d6c16f454b06829fc93483440916cc3cf0, but `w` is unused after this. Not sure if that's a problem or just a useless line tho' :)

hrydgard replied:
Looks to me like the line is unnecessary indeed.

unknownbrackets replied:
It's just hung around since here:
https://github.com/hrydgard/ppsspp/commit/5fb8e6dfb07ad11efa7ec91ec2e632d03825e34b#diff-06771fb6126492de2d06dd342494acf73e3206bf2ef7d15bda486c9365cb5944R847

-[Unknown]

hrydgard committed:
Remove redundant line

hrydgard committed:
Simplification and some cleanup

hrydgard committed:
Bugfixes, handle the rest of the cases

hrydgard committed:
Debug log (remove before merge)

hrydgard committed:
Yet another bugfix

hrydgard committed:
SSE optimize a common case for video

hrydgard commented:
OK, this is probably ready to go, barring any more silly bugs. Greatly speeds up decoding of large textures in Vulkan, though there's a bit more to eke out here and there.

unknownbrackets commented on GPU/Vulkan/TextureCacheVulkan.cpp:
```diff
@@ -980,18 +982,11 @@ void TextureCacheVulkan::LoadTextureLevel(TexCacheEntry &entry, uint8_t *writePt
+
+		CheckAlphaResult alphaResult = DecodeTextureLevel((u8 *)pixelData, decPitch, tfmt, clutformat, texaddr, level, bufw, false, false, expand32);
+
+		WARN_LOG(G3D, "Alpha: full=%d w=%d h=%d level=%d %s/%s", (int)(alphaResult == CHECKALPHA_FULL), w, h, level, GeTextureFormatToString(tfmt), GEPaletteFormatToString(clutformat));
```
Temporary logging, right?

-[Unknown]

unknownbrackets commented on GPU/Common/TextureDecoder.cpp:
```diff
@@ -610,10 +626,11 @@ uint32_t GetDXT5Texel(const DXT5Block *src, int x, int y) {
-	dxt.DecodeColors(src, ignore1bitAlpha);
+	dxt.DecodeColors(src, false);
 	dxt.WriteColorsDXT1(dst, src, pitch, height);
+	*alpha = dxt.AnyNonFullAlpha() ? 0 : 1;
```
nit: I realize we set the one bit, but seems like `-1` or `0xFFFFFFFF` might be clearer here.

-[Unknown]

unknownbrackets commented on GPU/Common/TextureCacheCommon.cpp:
```diff
@@ -1439,145 +1562,175 @@ void TextureCacheCommon::DecodeTextureLevel(u8 *out, int outPitch, GETextureForm
 			const u8 *unswizzled = (u8 *)tmpTexBuf32_.data();
 
 			if (reverseColors) {
+				fullAlphaMask = TfmtRawToFullAlpha(format);
 				for (int y = 0; y < h; ++y) {
+					fullAlphaMask = TfmtRawToFullAlpha(format);
```
This looks off - I guess we could CheckMask32 here?

-[Unknown]

hrydgard replied:
Yeah looks like I copy pasted the wrong line, hah

unknownbrackets commented on GPU/Common/TextureCacheCommon.cpp:
```diff
@@ -1439,145 +1562,175 @@ void TextureCacheCommon::DecodeTextureLevel(u8 *out, int outPitch, GETextureForm
 		if (!swizzled) {
 			// Just a simple copy, we swizzle the color format.
 			if (reverseColors) {
+				// Just check the input's alpha to reuse code. TODO: make a specialized ReverseColors that checks as we go.
+				fullAlphaMask = TfmtRawToFullAlpha(format);
```
Shouldn't this be above `if (reverseColors)`?  It's currently missing in the `expandTo32bit` case.

-[Unknown]

hrydgard replied:
Yup, fixing

unknownbrackets commented on GPU/Common/TextureCacheCommon.cpp:
```diff
@@ -1439,145 +1562,175 @@ void TextureCacheCommon::DecodeTextureLevel(u8 *out, int outPitch, GETextureForm
 			const u8 *unswizzled = (u8 *)tmpTexBuf32_.data();
 
 			if (reverseColors) {
+				// Just check the swizzled input's alpha to reuse code. TODO: make a specialized ReverseColors that checks as we go.
+				fullAlphaMask = TfmtRawToFullAlpha(format);
```
nit: Could move above `if (reverseColors) {` and get a bit more DRY.

-[Unknown]

hrydgard replied:
Indeed

hrydgard committed:
Quick NEON optimization of CheckMask32

hrydgard committed:
Fix some NEON code that had bad compile-time checks (and some didn't compile)

hrydgard committed:
One more NEON optimization

hrydgard committed:
Address feedback

hrydgard commented:
Thanks for looking, fixed the issues! 

This currently also includes https://github.com/hrydgard/ppsspp/pull/15481 as the first commit, rebased on that. But let's merge that one before this one.
